src/WRings.sol

```diff
@@ -0,0 +1,124 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.13;
+
+import "./Rings.sol";
+import "../lib/yield-utils-v2/contracts/mocks/ERC20Mock.sol";
+import "../lib/yield-utils-v2/contracts/token/IERC20.sol";
+import "../lib/yield-utils-v2/contracts/math/WDiv.sol";
+import "../lib/yield-utils-v2/contracts/math/WMul.sol";
+import "../lib/yield-utils-v2/contracts/access/Ownable.sol";
+
+/// @title WRings
+/// @author hashedMae
+/// @notice An ERC20 fractional wrapper for yield bearing tokens
+/// @dev Utilizes WMul and WDiv math libraries from Yield-Utils-V2. Will always round down to zero.
+
+contract WRings is ERC20Mock,
+                   Ownable {
+
+    IERC20 public immutable iRings;
+    uint256 public exchangeRate;
+
+    /// @notice Emitted whenever a user wraps tokens
+    event Deposit(address indexed caller, address indexed owner, uint256 assets, uint256 shares);
+    /// @notice Emitted whenever a user unwraps tokens
+    event Withdraw(address indexed caller, address indexed receiver, address indexed owner, uint256 assets, uint256 shares);
+
+    /// @param iRings_ ERC20 Interface for Rings token
+    constructor(IERC20 iRings_) ERC20Mock("WRings", "WRNG") {
+        iRings = iRings_;
+    }
+
+    /// @return assetTokenAddress Contract address of token utilized by the vault
+    function asset() public view returns(address){
+        address assetTokenAddress = address(iRings);
+        return assetTokenAddress;
+    }
+
+    /// @return totalManagedAssets Balance of the asset token within the vault
+    function totalAssets() public view returns(uint256) {
+        uint256 totalManagedAssets = iRings.balanceOf(address(this));
+        return totalManagedAssets;
+    }
+
+    /// @notice Allows the owner to set the exchange rate
+    /// @dev uses 27 decimals of precision, and can be no greater than 1000000000000000000
+    /// @param exchangeRate_ New exchange rate
+    function setExchangeRate(uint256 exchangeRate_) external onlyOwner {
+        require(exchangeRate_ <= 100000000000000000 * 10 ** 27, "New rate is outside bounds");
+        exchangeRate = exchangeRate_;
+    }
+
+    /// @notice The amount of shares to mint to a user for the asset token deposited under ideal conditions
+    /// @param assets Number of Rings tokens to deposit to the contract 
+    /// @return shares Number of vault shares minted to the user based on the exchange rate
+    function convertToShares(uint256 assets) external view returns(uint256) {
+        uint256 shares = WMul.wmul(assets * 10**9, exchangeRate) / 10**9;
```

Comment:
The `W` in `WMul` stands for `Wad`, which is MakerDAO-speak for "18 decimals". The other two decimal lengths are `ray` for 27 and `rad` for 45 decimals. See [here](https://github.com/makerdao/dss/blob/17187f7d47be2f4c71d218785e1155474bbafe8a/src/vat.sol#L44).

Instead of shoehorning WMul to work with rays, I suggest that you create your own RMath.sol library with `rmul` and and `rdiv` functions, import it here, and link it using `using` as in `using RMath for uint256;` at the top of the `Wrings` contract.

 Then this line would be `uint256 shares = assets.wmul(exchangeRate);`

Reply:
I could never figure out or find anything explaining what wad, ray, or rad meant or why they were used in so many places. Thank you for putting me in the know. 

Making a math library sounds like fun